presentation/json.go

```diff
@@ -0,0 +1,134 @@
+package presentation
+
+import (
+	"encoding/json"
+	"github.com/box/memsniff/analysis"
+	"os"
+	"time"
+)
+
+func (u *uiContext) runLoggingEventLoop() error {
+	updateTick := time.NewTicker(u.interval)
+	defer updateTick.Stop()
+	if err := u.update(); err != nil {
+		return err
+	}
+
+	for {
+		select {
+		case <-updateTick.C:
+			if err := u.updateReport(); err != nil {
+				return err
+			}
+
+		case msg := <-u.msgChan:
+			u.logger.Log(msg)
+		}
+	}
+}
+
+func (u *uiContext) updateReport() error {
+	rep := u.analysis.Report(!u.cumulative)
+	rep.SortBy(-2)
+
+	numKeysSeen := len(rep.Rows)
+	totalBandwidthUsed := totalBytesUseForKeys(rep.Rows)
+
+	s := u.statProvider()
+
+	u.truncateResultsToMaxAndTopX(&rep)
+	u.prevReport = rep
+	reportedKeysBandwidth := totalBytesUseForKeys(rep.Rows)
+
+	var f *os.File = nil
+	if u.outputFile != "" {
+		var err error
+		f, err = os.OpenFile(u.outputFile, os.O_RDWR|os.O_CREATE|os.O_APPEND|os.O_SYNC, 0644)
+		if err != nil {
+			return err
+		}
+		defer f.Close()
+	}
+
+	reportJson := formatReportAsJson(u.prevReport, s, numKeysSeen, totalBandwidthUsed, reportedKeysBandwidth)
+	reportJsonBytes, err := json.Marshal(reportJson)
+	reportJsonString := string(reportJsonBytes)
+
+	if err != nil {
+		u.Log(err)
+	} else {
+		if f != nil {
+			// Write to the output file if specified
+			f.WriteString(reportJsonString + "\n")
+		} else {
+			u.Log(reportJsonString)
+		}
+	}
+
+	return nil
+}
+
+type JsonReport struct {
+	Timestamp                   string
+	TotalKeys                   int
+	TotalBandwidth              int64
+	ReportedKeys                int
+	ReportedBandwidth           int64
+	ReportedBandwidthPercentage float64
+	Rows                        []map[string]interface{}
+	Stats                       StatsSet
+}
+
+func formatReportAsJson(report analysis.Report, stats StatsSet, totalKeys int, totalBandwidth int64, reportedBandwidth int64) JsonReport {
+	var reportedBandwidthPercentage float64 = 0
+	if reportedBandwidth > 0 && totalBandwidth > 0 {
+		reportedBandwidthPercentage = 100 * float64(reportedBandwidth) / float64(totalBandwidth)
+	}
+
+	return JsonReport{
+		Timestamp:                   report.Timestamp.Format("2006-01-02T15:04:05-0700"),
```

Comment:
There is a predefined constant: time.RFC3339 (https://golang.org/pkg/time/#pkg-constants)